File: javascript/leetcode/922. Sort Array By Parity II.js
/**
 * @param {number[]} A
 * @return {number[]}
 */
var sortArrayByParityII = function(A) {
    let odd = 1
    
    for(let i = 0; i < A.length; i += 2) { 
        if(A[i] % 2 !== 0) {
            while(A[odd] % 2 !== 0) 
                odd += 2
            
            const temp = A[i]
            A[i] = A[odd]
            A[odd] = temp 
        } 
    }
    
    return A
}; 

File: c++/leetcode/922. Sort Array By Parity II.cpp
class Solution {
public:
    vector<int> sortArrayByParityII(vector<int>& A) {
        queue<int> odd;
        queue<int> even;
        
        for(auto i : A) {
            if(i % 2 == 0) 
                even.push(i);
            else
                odd.push(i);
        }
        
        vector<int> ans;
        
        for(int i = 0; i < A.size(); ++i) { 
            if(i % 2 == 0) {
                ans.push_back(even.front());
                even.pop();
            } else {
                ans.push_back(odd.front());
                odd.pop();
            }
        }        
        
        return ans;
    }
};
